identity/src/main/java/com/android/identity/L2CAPClient.java

```diff
@@ -0,0 +1,234 @@
+    private boolean mInhibitCallbacks = false;
+
+    L2CAPClient(@Nullable Listener listener, @LoggingFlag int loggingFlags) {
+        mListener = listener;
+        mLog = new Util.Logger(TAG, loggingFlags);
+    }
+
+    void disconnect() {
+        mInhibitCallbacks = true;
+        try {
+            if (mSocket != null) {
+                mSocket.close();
+                mSocket = null;
+            }
+        } catch (IOException e) {
+            // Ignoring this error
+            Log.e(TAG, " Error closing socket connection " + e.getMessage(), e);
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.Q)
+    public void connect(@NonNull BluetoothDevice bluetoothDevice, byte[] psmValue) {
+        if (psmValue == null || psmValue.length == 0 || psmValue.length > 4) {
+            reportError(new Error("Invalid PSM value received on L2CAP characteristic"));
+            return;
+        }
+        byte[] psmSized = new byte[4];
+        if (psmValue.length < 4) {
+            // Add 00 on left if psm length is lower than 4
+            System.arraycopy(psmValue, 0, psmSized, 4 - psmValue.length, psmValue.length);
+        } else {
+            psmSized = psmValue;
+        }
+        int psm = ByteBuffer.wrap(psmSized).getInt();
+        if (mLog.isTransportEnabled()) {
+            mLog.transport("Received psmValue: " + Util.toHex(psmValue) + " psm: " + psm);
+        }
+
+        try {
+            mSocket = bluetoothDevice.createInsecureL2capChannel(psm);
+            mSocket.connect();
+            if (isConnected()) {
+                mLog.transport("Connected using L2CAP on PSM: " + psm);
+                Thread writingThread = new Thread(this::writeToSocket);
+                writingThread.start();
+
+                Thread readingThread = new Thread(this::readFromSocket);
+                readingThread.start();
+
+                reportPeerConnected();
+            } else {
+                reportError(new Error("Unable to connect L2CAP socket"));
+            }
+        } catch (IOException e) {
+            Log.e(TAG, "Error connecting to socket L2CAP " + e.getMessage(), e);
+            reportError(new Error("Error connecting to socket L2CAP", e));
+        }
+
+    }
+
+    private void writeToSocket() {
+        OutputStream os;
+
+        try {
+            os = mSocket.getOutputStream();
+        } catch (IOException e) {
+            Log.e(TAG, "Error writing message on L2CAP socket", e);
+            reportError(new Error("Error writing message on L2CAP socket", e));
+            return;
+        }
+
+        if (mLog.isTransportEnabled()) {
+            mLog.transport("Writing L2CAP socket");
+        }
+        while (isConnected()) {
+            byte[] messageToSend;
+            try {
+                messageToSend = mWriterQueue.poll(1000, TimeUnit.MILLISECONDS);
+                if (messageToSend == null) {
+                    continue;
+                }
```

Comment:
Check if messageToSend.length is 0, and if so, break out of the loop to end the writing thread just like in https://github.com/google/identity-credential/blob/master/identity/src/main/java/com/android/identity/DataTransportTcp.java#L261